internal/mq/msgstream/mqwrapper/kafka/kafka_consumer.go

```diff
@@ -80,111 +95,76 @@ func (kc *Consumer) Subscription() string {
 // channel-based consumer API had already deprecated, see more details
 // https://github.com/confluentinc/confluent-kafka-go.
 func (kc *Consumer) Chan() <-chan mqwrapper.Message {
+	if kc.hasAssign != true {
+		log.Error("can not chan with not assigned channel", zap.String("topic", kc.topic), zap.String("groupID", kc.groupID))
+		panic("failed to chan a kafka consumer without assign")
+	}
 	kc.chanOnce.Do(func() {
-		if !kc.hasSeek {
-			offsetStr, err := kc.config.Get("auto.offset.reset", "earliest")
-			if err != nil {
-				log.Error("get auto.offset.reset config fail in kafka consumer", zap.String("topic name", kc.topic), zap.Error(err))
-				panic(err)
-			}
-
-			offset, err := kafka.NewOffset(offsetStr)
-			if err != nil {
-				log.Error("Invalid kafka offset", zap.String("topic name", kc.topic), zap.Error(err))
-				panic(err)
-			}
-
-			// we assume that case is Chan starting before producing message with auto create topic config,
-			// consuming messages will fail that error is 'Subscribed topic not available'
-			// if invoke Subscribe method of kafka, so we use Assign instead of Subscribe.
-			var tps []kafka.TopicPartition
-			if offset == kafka.OffsetEnd && kc.latestMsgOffset != kafka.OffsetBeginning {
-				// kafka consumer will start when assign invoked, in order to guarantee the latest message
-				// position is same with created consumer time, there will use a seek to the latest to
-				// replace consuming from the latest position.
-				if err := kc.internalSeek(kc.latestMsgOffset, false); err != nil {
-					log.Error("kafka consumer subscribe failed ",
-						zap.String("topic name", kc.topic),
-						zap.Any("latestMsgOffset", kc.latestMsgOffset),
-						zap.Any("offset", offset),
-						zap.Error(err))
-					panic(err)
-				}
-			} else {
-				tps = []kafka.TopicPartition{{Topic: &kc.topic, Partition: mqwrapper.DefaultPartitionIdx, Offset: offset}}
-				if err := kc.c.Assign(tps); err != nil {
-					log.Error("kafka consumer subscribe failed ",
-						zap.String("topic name", kc.topic),
-						zap.Any("latestMsgOffset", kc.latestMsgOffset),
-						zap.Any("offset", offset),
-						zap.Error(err))
-					panic(err)
-				}
-			}
-
-			log.Debug("starting kafka consume",
-				zap.String("topic name", kc.topic),
-				zap.Any("latestMsgOffset", kc.latestMsgOffset),
-				zap.Any("offset", offset))
-		}
-
 		go func() {
-			// loop end if consumer is closed
-			for ev := range kc.c.Events() {
-				switch e := ev.(type) {
-				case *kafka.Message:
-					if kc.skipMsg {
-						kc.skipMsg = false
-						continue
+			for {
+				select {
+				case _ = <-kc.closeCh:
+					log.Info("close consumer ", zap.String("topic", kc.topic), zap.String("groupID", kc.groupID))
+					start := time.Now()
+					err := kc.c.Unassign()
```

Comment:
It seems that we don't need to unassign because the close consumer will unsubscribe.

Reply:
removed